File: week5_dynamic_programming1/4_longest_common_subsequence_of_two_sequences/lcs2.py
# Uses python3

import sys


def lcs2(a, b):
    # write your code here
    result = [[0 for j in range(m+1)] for i in range(n + 1)]
    for x in range(1, n + 1):
        for y in range(1, m + 1):
            if a[x - 1] == b[y - 1]:
                result[x][y] = result[x - 1][y - 1] + 1
            else:
                result[x][y] = max(result[x - 1][y], result[x][y - 1])
    return result


if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))

    n = data[0]
    data = data[1:]
    a = data[:n]

    data = data[n:]
    m = data[0]
    data = data[1:]
    b = data[:m]
    result = lcs2(a, b)
    print(result[n][m])

File: week5_dynamic_programming1/4_longest_common_subsequence_of_two_sequences/lcs2.cpp
#include <iostream>
#include <vector>

using std::vector;


int lcs2(vector<int> &a, vector<int> &b) {
  //write your code here
  vector<int> result;
  int m = a.size();
  int n = b.size();
  int t = max(m,n);
  int D[m+1][n+1];
  for(int i=0;i<=m;++i) D[i][0]=i;
  for(int i=0;i<=n;++i) D[0][i]=i;
  for(int i=1;i<=m;++i)
  	for(int j=1;j<=n;++j)
  	{
  		int op[3]={t,t,t}; //t is equal to infinite 
  		op[0] = D[i-1][j]+1;
  		op[1] = D[i][j-1]+1;
  		if(a[i-1]==b[j-1]) op[2] = D[i-1][j-1];
  		else op[2] = D[i-1][j-1] + 1; 
  		D[i][j] = min(op[1],op[0]);
  		D[i][j] = min(D[i][j],op[2]);
	}
  	//Backward
  	int i = m;
  	int j = n;
  	while(i!=0&&j!=0)
  	{
  		if(D[i][j]==D[i-1][j-1]&&D[i][j]<=D[i][j-1]&&D[i][j]<=D[i-1][j]) 
		{
		  result.push_back(a[j-1]);
		  i--;
		  j--;
		  continue;	
		}
		if(D[i-1][j]<=D[i][j-1])
		{
			i--;
			continue;
		}
		if(D[i][j-1]<=D[i-1][j])
		{
			j--;
			continue;
		}
	}
  return result.size();
}




// int lcs2(vector<int> &a, vector<int> &b) {
//   //write your code here
//   return std::min(std::min(a.size(), b.size()), c.size());
// }

int main() {
  size_t n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < n; i++) {
    std::cin >> a[i];
  }

  size_t m;
  std::cin >> m;
  vector<int> b(m);
  for (size_t i = 0; i < m; i++) {
    std::cin >> b[i];
  }

  std::cout << lcs2(a, b) << std::endl;
}
